scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java

```diff
@@ -12,125 +13,477 @@
+        for (ColumnDefinitions.Definition cdcColumn : cdcColumnDefinitions) {
+            String columnName = cdcColumn.getName();
+            ChangeSchema.DataType cdcDataType = translateCdcColumnDataType(cdcColumn.getType());
+
+            ChangeSchema.DataType baseTableDataType = null;
+            ChangeSchema.ColumnType baseTableColumnType = null;
+            if (!columnName.startsWith("cdc$")) {
+                baseTableDataType = inferBaseTableDataType(cdcColumn, cdcDataType);
+                baseTableColumnType = baseTableSchema.getColumnType(columnName, cdcColumnDefinitions);
+            }
+
+            builder.add(new ChangeSchema.ColumnDefinition(columnName, index++, cdcDataType, baseTableDataType, baseTableColumnType));
         }
-        return builder.build();
+
+        return new ChangeSchema(builder.build());
     }
 
-    private ChangeSchema.ColumnDefinition translateColumnDefinition(ColumnDefinitions.Definition driverDefinition, int index) {
-        String columnName = driverDefinition.getName();
-        ChangeSchema.DataType dataType = translateColumnDataType(driverDefinition.getType());
-        ChangeSchema.DataType baseType = null;
-        ChangeSchema.ColumnType baseTableColumnType = ChangeSchema.ColumnType.REGULAR;
-        if (baseTablePartitionKeyColumnNames.contains(columnName)) {
-            baseTableColumnType = ChangeSchema.ColumnType.PARTITION_KEY;
-        } else if (baseTableClusteringKeyColumnNames.contains(columnName)) {
-            baseTableColumnType = ChangeSchema.ColumnType.CLUSTERING_KEY;
+    private ChangeSchema.DataType inferBaseTableDataType(ColumnDefinitions.Definition cdcColumn, ChangeSchema.DataType cdcDataType) throws UnresolvableSchemaInconsistencyException {
+        if (!cdcDataType.isNative()) {
+            // In the CDC table we have a frozen non-native type (frozen collection,
+            // frozen UDT or frozen tuple). That means that in the base table there
+            // is a same type - but it could be frozen or non-frozen.
+            // We use the existence of cdc$deleted_elements_ column as a proxy
+            // for that information.
+
+            boolean hasDeletedElementsColumn =
+                    cdcColumnDefinitions.contains("cdc$deleted_elements_" + cdcColumn.getName());
+            if (hasDeletedElementsColumn) {
+                // If the column has cdc$deleted_elements_,
+                // that means it was non-frozen in base table.
+
+                // LIST<X>, MAP<TIMEUUID, X> in the base table
+                // are both represented as FROZEN<MAP<TIMEUUID, X>>
+                // in the CDC log table. This checks if it
+                // was a LIST<X> in the base table.
+                if (cdcDataType.getCqlType() == ChangeSchema.CqlType.MAP
+                        && cdcDataType.getTypeArguments().get(0).getCqlType() == ChangeSchema.CqlType.TIMEUUID
+                        && baseTableSchema.isList(cdcColumn.getName())) {
```

Comment:
The precondition to `isList` (and this precondition is already documented!) is that we know from the CDC schema that this column is a `list<X>` or `map<timeuuid, X>` in the base table schema (as of CDC log schema time).

`isList` throws inconsistency exception in a case it encounters a different type than `list<X>` or `map<timeuuid, X>`. This is for a case when someone changes a type from `list<X>` to `blob`, the CDC schema is old (`frozen<map<timeuuid, X>>`) and base table schema is fresh (`blob`). From such information you cannot determine if the type was `list<X>` or `map<timeuuid, X>` at the time of CDC schema.

If you'd remove the first and second conjuncts, it would throw an exception for each non-list, non-map-timeuuid collection type. Of course, you could move this logic (first and second conjuncts) from this place into `isList` - maybe a valid suggestion.

> Is it even possible to change non-frozen collection columns' types?

Yes, to `blob` (although I don't see such type change very useful...)